packages/vm/src/modules/file_system_cache.rs

```diff
@@ -62,14 +62,23 @@ impl FileSystemCache {
         }
     }
 
-    /// Loads an artifact from the file system and returns a module (i.e. artifact + store).
-    pub fn load(&self, checksum: &Checksum, store: &Store) -> VmResult<Option<Module>> {
+    /// Loads an artifact from the file system and returns a module (i.e. artifact + store),
+    /// along with the artifact size.
+    pub fn load(&self, checksum: &Checksum, store: &Store) -> VmResult<Option<(Module, usize)>> {
         let filename = checksum.to_hex();
         let file_path = self.latest_modules_path().join(filename);
 
         let result = unsafe { Module::deserialize_from_file(store, &file_path) };
```

Comment:
I was just thinking that if we make `deserialize_from_file` / `serialize_from_file` return the serialized module size, we can dispose of the `metadata()` calls, here and in `store()`.

Do I go ahead and modify `wasmer` / create an issue in wasmer to modify this? I think it will make all the code / logic really elegant.

Reply:
This sounds a bit out of scope for the Wasmer interface. As long as we use the JIT engine only, we can read the file into a vector ourselves and then use `Module::deserialize`. Then we have the size without another call. However, if we move to the native engine, `Module::deserialize` becomes a problem as it must write the bytes to file in order to be able to use dlopen.

> Using this, we already got rid of the `OPTIMIZED_WASM_SIZE_FACTOR` kludge, by the way.

Oh, nice

Reply:
In any case, here it is: https://github.com/wasmerio/wasmer/issues/2055

Reply:
> Then we have the size without another call.

My point here's that they already have this information. Moreover: these two functions both return `Result<()>` so, it's straightforward (and mostly non-breaking) to return the size instead.

Reply:
> As long as we use the JIT engine only, we can read the file into a vector ourselves and then use `Module::deserialize`

I can do that in the mean time (or in case they didn't agree on implementing this), if you think it's worth it.

Reply:
Or, we can leave the extra `metadata()` calls.

Reply:
I think there is nothing wrong with the metadata call.